> ### Note on Labs and Assigments:
>
> 🔧 Look for the **wrench emoji** 🔧 — it highlights where you're expected to take action!
>
> These sections are graded and are not optional.
>

# IS 4487 Lab 6: Data Cleaning

## Outline

- Load and inspect a new dataset (Megatelco)
- Fix column names and data types
- Handle missing values
- Remove duplicate rows
- Review and remove outliers
- Reflect on data quality

In this lab, we’ll clean the data to get it ready for transformations and analysis.

We will continue working with this dataset in **Lab 7**, where we will create new features and apply transformations.

<a href="https://colab.research.google.com/github/Stan-Pugsley/is_4487_base/blob/main/Labs/lab_06_data_cleaning.ipynb" target="_parent">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

## Megatelco Data Dictionary

 DEMOGRAPHIC VARIABLES:
 - College - has the customer attended some college (one, zero)
 - Income - annual income of customer
 - House - estimated price of the customer's home (if applicable)

 USAGE VARIABLES:
 - Data Overage Mb - Average number of megabytes that the customer used in excess of the plan limit (over last 12 months)
 - Data Leftover Mb - Average number of megabytes that the customer use was below the plan limit (over last 12 months)
 - Data Mb Used - Average number of megabytes used per month (over last 12 months)
 - Text Message Count - Average number of texts per month (over last 12 months)
 - Over 15 Minute Calls Per Month - Average number of calls over 15 minutes in duration per month (over last 12 months)
 - Average Call Duration- Average call duration (over last 12 months)

PHONE VARIABLES:
 - Operating System - Current operating system of phone
 - Handset Price - Retail price of the phone used by the customer

ATTITUDINAL VARIABLES:
 - Reported Satisfaction - Survey response to "How satisfied are you with your current phone plan?" (high, med, low)
 - Reported Usage Level - Survey response to "How much do your use your phone?" (high, med, low)
 - Considering Change of Plan - Survey response to "Are you currently planning to change companies when your contract expires?" (high, med, low)

OTHER VARIABLES
 - Leave - Did this customer churn with the last contract expiration? (LEAVE, STAY)
 - ID - Customer identifier

In [14]:
import pandas as pd

url = "https://raw.githubusercontent.com/Stan-Pugsley/is_4487_base/812e9f15c357a5657a2795631fcaa9d9363cb417/DataSets/megatelco_leave_survey_data_cleaning_v2.csv"
df = pd.read_csv(url)

df.head()

college    income  data_overage_mb  data_leftover_mb  data_mb_used  \
0     one  403137.0               70               0.0        6605.0   
1    zero  129700.0               67              16.0        6028.0   
2    zero   69741.0               60               0.0        1482.0   
3     one  377572.0                0              22.0        3005.0   
4    zero  382080.0                0               0.0        1794.0   

   text_message_count   house  handset_price  over_15mins_calls_per_month  \
0                 199  841317          653.0                          5.0   
1                 134  476664         1193.0                          5.0   
2                 176  810225         1037.0                          3.0   
3                 184  826967         1161.0                          0.0   
4                  74  951896         1023.0                          0.0   

   average_call_duration reported_satisfaction reported_usage_level  \
0                    8.0                   low                  low   
1                    5.0                   low                  low   
2                    8.0                   low                  low   
3                    5.0                   low                  low   
4                   14.0                   low                  low   

  considering_change_of_plan  leave     id operating_system  
0                        yes  LEAVE   8183          Android  
1                        yes  LEAVE  12501              IOS  
2                        yes   STAY   7425              IOS  
3                         no  LEAVE  13488              IOS  
4                        yes   STAY  11389              IOS

In [15]:
# create a copy of your dataset for use in part 4
copied_df = df.copy(deep=True)

## Part 1: Review Column Names and Structure

Before cleaning, check the structure of the dataset:

- Are column names consistent (lowercase, no spaces)?
- Are there any typos or redundant labels?
- Do the rows and columns appear aligned? (Are all the columns the same size? Are all the rows the same size?)

Why this matters:
Inconsistent or messy column names can break code and make analysis harder to follow.




In [16]:
# Standardize column names: lowercase, no spaces
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_")

# Get column info and data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15016 entries, 0 to 15015
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   college                      15016 non-null  object 
 1   income                       15006 non-null  float64
 2   data_overage_mb              15016 non-null  int64  
 3   data_leftover_mb             14916 non-null  float64
 4   data_mb_used                 14916 non-null  float64
 5   text_message_count           15016 non-null  int64  
 6   house                        15016 non-null  int64  
 7   handset_price                14916 non-null  float64
 8   over_15mins_calls_per_month  15013 non-null  float64
 9   average_call_duration        14916 non-null  float64
 10  reported_satisfaction        15016 non-null  object 
 11  reported_usage_level         15016 non-null  object 
 12  considering_change_of_plan   14201 non-null  object 
 13  leave           

## Part 2: Convert Data Types

Before analysis, make sure each column is stored in the correct format. This helps avoid calculation errors, makes plotting smoother, and ensures models interpret the data correctly.

Think about:
- Are numbers accidentally stored as strings?
- Should repeated text values be converted to categories?
- Are "yes"/"no" columns better represented as binary (0/1) or categorical types?

Fixing data types now saves time and avoids issues later in your workflow.




In [17]:
# Check original data types
print("Original dtypes:\n", df.dtypes)

# Convert yes/no to binary categories
df['considering_change_of_plan'] = df['considering_change_of_plan'].map({'yes': 1, 'no': 0}).astype('category')

# Convert categorical text columns
df['college'] = df['college'].astype('category')
df['house'] = df['house'].astype('category')
df['reported_satisfaction'] = df['reported_satisfaction'].astype('category')
df['operating_system'] = df['operating_system'].astype('category')

# Convert numeric-looking columns from object to float
df['income'] = pd.to_numeric(df['income'], errors='coerce')
df['data_overage_mb'] = pd.to_numeric(df['data_overage_mb'], errors='coerce')
df['data_leftover_mb'] = pd.to_numeric(df['data_leftover_mb'], errors='coerce')
df['data_mb_used'] = pd.to_numeric(df['data_mb_used'], errors='coerce')
df['text_message_count'] = pd.to_numeric(df['text_message_count'], errors='coerce')
df['over_15mins_calls_per_month'] = pd.to_numeric(df['over_15mins_calls_per_month'], errors='coerce')
df['average_call_duration'] = pd.to_numeric(df['average_call_duration'], errors='coerce')

# Check updated data types
print("\nUpdated dtypes:\n", df.dtypes)


Original dtypes:
 college                         object
income                         float64
data_overage_mb                  int64
data_leftover_mb               float64
data_mb_used                   float64
text_message_count               int64
house                            int64
handset_price                  float64
over_15mins_calls_per_month    float64
average_call_duration          float64
reported_satisfaction           object
reported_usage_level            object
considering_change_of_plan      object
leave                           object
id                               int64
operating_system                object
dtype: object

Updated dtypes:
 college                        category
income                          float64
data_overage_mb                   int64
data_leftover_mb                float64
data_mb_used                    float64
text_message_count                int64
house                          category
handset_price                   float64
over_1

### 🔧 Try It Yourself – Part 2

1. Convert the `leave` column from "yes"/"no" to binary (`1`/`0`) and make it a **category**
2. Convert `reported_usage_level` to a **categorical** type
3. Convert `house` to an **integer** type
3. Use `.info()` to confirm the changes


In [18]:
# Convert 'leave' column from "yes"/"no" to binary (1/0) and make it a category
df['leave'] = df['leave'].map({'LEAVE': 1, 'STAY': 0}).astype('category')

# Convert 'reported_usage_level' to a categorical type
df['reported_usage_level'] = df['reported_usage_level'].astype('category')

# Convert 'house' to an integer type
# First, convert 'house' back to numeric, then fill any potential remaining NaNs with 0 as a safe default before converting to int
df['house'] = pd.to_numeric(df['house'], errors='coerce').fillna(0).astype(int)


# Use .info() to confirm the changes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15016 entries, 0 to 15015
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype   
---  ------                       --------------  -----   
 0   college                      15016 non-null  category
 1   income                       15006 non-null  float64 
 2   data_overage_mb              15016 non-null  int64   
 3   data_leftover_mb             14916 non-null  float64 
 4   data_mb_used                 14916 non-null  float64 
 5   text_message_count           15016 non-null  int64   
 6   house                        15016 non-null  int64   
 7   handset_price                14916 non-null  float64 
 8   over_15mins_calls_per_month  15013 non-null  float64 
 9   average_call_duration        14916 non-null  float64 
 10  reported_satisfaction        15016 non-null  category
 11  reported_usage_level         15016 non-null  category
 12  considering_change_of_plan   14201 non-null  category
 13  l

## Part 3: Handle Missing Values

Missing data can break charts, skew stats, and disrupt models — so it needs to be handled carefully.

### Think about:
- Are the missing values random or patterned?
- Can we drop rows, or do we need to fill them?
- Should we use mean, median, or something else?

### Guidelines:
- Drop rows if there are only a few missing and the columns associated with them are essential to keep intact
- Use median to replace outliers in numeric columns
- Use 0 if the missing value means “none” (e.g. If the value was in response to: “do you have a history of chronic illness?” and the value was just left blank, we can assume that that blank just means “none” (the patient has no history of chronic illness))
- Use mode to replace categorical values

Cleaning missing values early avoids bigger problems later.

-----


**Note on `.loc` and Warnings** - When assigning values to a DataFrame, especially after filtering or copying, it's best to use `.loc` to avoid **`SettingWithCopyWarning`**. This ensures that you're updating the original data and not a temporary view of it.


In [19]:
# View missing value counts
print("Missing values per column:\n", df.isnull().sum())

# Fill 'handset_price' with median
df['handset_price'] = df['handset_price'].fillna(df['handset_price'].median())

# Drop rows with missing 'income' (if very few)
df = df.dropna(subset=['income']).copy()

# Fill missing 'data_leftover_mb' with 0 if it logically means no leftover data
df.loc[:, 'data_leftover_mb'] = df['data_leftover_mb'].fillna(0)

# Fill 'average_call_duration' with median if necessary
df.loc[:, 'average_call_duration'] = df['average_call_duration'].fillna(df['average_call_duration'].median())

# Fill 'data_mb_used' with median
df.loc[:, 'data_mb_used'] = df['data_mb_used'].fillna(df['data_mb_used'].median())

# Confirm updated missing values
print("\nMissing values after handling:\n", df.isnull().sum())


Missing values per column:
 college                          0
income                          10
data_overage_mb                  0
data_leftover_mb               100
data_mb_used                   100
text_message_count               0
house                            0
handset_price                  100
over_15mins_calls_per_month      3
average_call_duration          100
reported_satisfaction            0
reported_usage_level             0
considering_change_of_plan     815
leave                            0
id                               0
operating_system                 0
dtype: int64

Missing values after handling:
 college                          0
income                           0
data_overage_mb                  0
data_leftover_mb                 0
data_mb_used                     0
text_message_count               0
house                            0
handset_price                    0
over_15mins_calls_per_month      3
average_call_duration            0
reported_satisfa

### 🔧 Try It Yourself – Part 3


There are still some missing values in:

- `over_15mins_calls_per_month`
- `considering_change_of_plan`

Decide how to handle them based on what makes the most sense:

- Should you fill them with 0, the median, or something else?
- For categories, would a placeholder like "unknown" or the most common value work?
- Or is it better to drop those rows?

1. Write and execute code to handle the missing values in the remaining two columns.
2. Use `df.isnull().sum()` to confirm all missing values are handled.



In [21]:
# Show the median of over_15mins_calls_per_month
over_15mins_median = df['over_15mins_calls_per_month'].median()
print(f"Median of over_15mins_calls_per_month: {over_15mins_median}")

# Show the mode of considering_change_of_plan and its frequency counts
considering_mode = df['considering_change_of_plan'].mode()[0]
print(f"\nMode of considering_change_of_plan: {considering_mode}")
print("\nFrequency counts for considering_change_of_plan:")
print(df['considering_change_of_plan'].value_counts(dropna=False))

# Fill missing values in over_15mins_calls_per_month with its median
df['over_15mins_calls_per_month'] = df['over_15mins_calls_per_month'].fillna(over_15mins_median)

# Fill missing values in considering_change_of_plan with its mode
df['considering_change_of_plan'] = df['considering_change_of_plan'].fillna(considering_mode)

# Confirm all missing values are handled
print("\nMissing values after handling:\n", df.isnull().sum())

Median of over_15mins_calls_per_month: 9.0

Mode of considering_change_of_plan: 1.0

Frequency counts for considering_change_of_plan:
considering_change_of_plan
1.0    10075
0.0     4931
Name: count, dtype: int64

Missing values after handling:
 college                        0
income                         0
data_overage_mb                0
data_leftover_mb               0
data_mb_used                   0
text_message_count             0
house                          0
handset_price                  0
over_15mins_calls_per_month    0
average_call_duration          0
reported_satisfaction          0
reported_usage_level           0
considering_change_of_plan     0
leave                          0
id                             0
operating_system               0
dtype: int64


## Part 4: Remove Duplicate Rows

Sometimes the same row appears more than once due to data entry or processing mistakes. It's important to check for and remove these duplicates.

Think about:
- Are there rows that are exactly the same?
- If duplicates exist, should you keep the first one, the last one, or none?

Why this matters:
Duplicate rows can inflate totals, distort statistics, and lead to inaccurate conclusions.


In [22]:
# Check for exact duplicates
print(f"Number of duplicate rows: {df.duplicated().sum()}")

# Remove them, keeping the first occurrence
df = df.drop_duplicates()

# Confirm result
print(f"Remaining rows after removing duplicates: {len(df)}")

Number of duplicate rows: 17
Remaining rows after removing duplicates: 14989


### 🔧 Try It Yourself – Part 4

1. Use `copied_df.duplicated().sum()` to count how many duplicates are in your dataset.
2. Try using `copied_df.drop_duplicates(keep='last')` instead — what changes?
3. Explore whether duplicate rows share the same ID or just values across all columns and comment on your observation.


In [24]:
# Count duplicates in copied_df
print(f"Number of duplicate rows in copied_df: {copied_df.duplicated().sum()}")

# Try dropping duplicates keeping the last occurrence
copied_df_last_duplicates = copied_df.drop_duplicates(keep='last')
print(f"Remaining rows in copied_df after dropping duplicates (keeping last): {len(copied_df_last_duplicates)}")

# Explore if duplicate rows share the same ID
# Find the duplicate rows
duplicate_rows = copied_df[copied_df.duplicated(keep=False)]

# Check if any duplicate rows have the same ID (this will check for rows that are duplicates in all columns *including* id)
print(f"\nNumber of duplicate rows that also have duplicate IDs: {duplicate_rows.duplicated(subset=['id']).sum()}")

# Alternatively, check if there are any duplicate IDs in the original copied_df
print(f"Number of duplicate IDs in copied_df: {copied_df['id'].duplicated().sum()}")

Number of duplicate rows in copied_df: 17
Remaining rows in copied_df after dropping duplicates (keeping last): 14999

Number of duplicate rows that also have duplicate IDs: 17
Number of duplicate IDs in copied_df: 3456


There were 17 exact duplicate rows tied to duplicate IDs. In addition, 3,456 IDs appear more than once, meaning many customers have multiple records in the dataset.

## Part 5: Identify and Remove Obvious Outliers

Outliers are values that fall far outside the normal range. They can come from data entry mistakes or rare cases.

- Use summary statistics or visual tools (like boxplots) to find them.
- Look for clearly unrealistic values — e.g., negative prices or extremely high data usage.
- Decide how to handle them:
  - Remove if they’re errors.
  - Keep if they’re valid but rare — or cap them if needed.

Outliers can distort averages, stretch visualizations, and mislead models, so it’s important to address them carefully.



In [25]:
# Remove negative or nonsensical values using business rules

# Example: remove rows where 'handset_price' is negative
df = df[df['handset_price'] >= 0]

# Example: remove rows with unusually long call durations
df = df[df['average_call_duration'] < 1000]

# Example: remove rows with extremely high text message counts
df = df[df['text_message_count'] < 1000]

# View shape after outlier filtering
print("Shape after removing obvious outliers:", df.shape)


Shape after removing obvious outliers: (14986, 16)


### 🔧 Try It Yourself – Part 5

1. Use `df.describe()` to look for columns with extreme minimum or maximum values.
2. Set a threshold for what you think is "too high" or "too low" for:
  - `data_mb_used`
  - `over_15mins_calls_per_month`
  - `income`
3. Remove those outliers using boolean filtering like `df = df[df['column'] < threshold]`

In [28]:
# Use df.describe() to look for columns with extreme minimum or maximum values.
print("Descriptive statistics before removing outliers:")
display(df.describe())

# Set thresholds and remove outliers using boolean filtering
# data_mb_used: too low if < 1000, too high if > 7000
df = df[(df['data_mb_used'] >= 1000) & (df['data_mb_used'] <= 7000)]

# over_15mins_calls_per_month: too low if < 1, too high if > 25
df = df[(df['over_15mins_calls_per_month'] >= 1) & (df['over_15mins_calls_per_month'] <= 25)]

# income: too low if < 0, too high if > 400000
df = df[(df['income'] >= 0) & (df['income'] <= 400000)]

# View shape after outlier filtering
print("\nShape after removing outliers:", df.shape)
print("\nDescriptive statistics after removing outliers:")
display(df.describe())

Descriptive statistics before removing outliers:


income  data_overage_mb  data_leftover_mb  data_mb_used  \
count    9354.000000      9354.000000       9354.000000   9354.000000   
mean   226190.368612       159.193286         37.683985   3990.370002   
std    100519.850725       110.789272         27.859459   1725.193341   
min     52491.000000         0.000000          0.000000   1001.000000   
25%    138562.000000        61.000000         13.000000   2497.000000   
50%    225878.500000       155.000000         35.000000   4044.000000   
75%    313180.000000       246.000000         62.000000   5487.750000   
max    399890.000000       380.000000         89.000000   7000.000000   

       text_message_count         house  handset_price  \
count         9354.000000  9.354000e+03    9354.000000   
mean           135.830874  8.735713e+05     794.461941   
std             48.901110  2.868228e+05    1325.558148   
min             52.000000  3.200010e+05     215.000000   
25%             93.000000  6.390412e+05     497.000000   
50%            136.000000  8.722050e+05     777.000000   
75%            178.000000  1.094315e+06    1065.000000   
max            220.000000  1.456389e+06  125000.000000   

       over_15mins_calls_per_month  average_call_duration            id  
count                  9354.000000            9354.000000   9354.000000  
mean                     10.480864               9.995724  11896.042228  
std                       6.992176               5.167853   6884.912231  
min                       1.000000               1.000000      2.000000  
25%                       4.000000               6.000000   6063.000000  
50%                      10.000000              10.000000  11796.000000  
75%                      17.000000              14.000000  17480.750000  
max                      25.000000              19.000000  25354.000000


Shape after removing outliers: (9354, 16)

Descriptive statistics after removing outliers:


income  data_overage_mb  data_leftover_mb  data_mb_used  \
count    9354.000000      9354.000000       9354.000000   9354.000000   
mean   226190.368612       159.193286         37.683985   3990.370002   
std    100519.850725       110.789272         27.859459   1725.193341   
min     52491.000000         0.000000          0.000000   1001.000000   
25%    138562.000000        61.000000         13.000000   2497.000000   
50%    225878.500000       155.000000         35.000000   4044.000000   
75%    313180.000000       246.000000         62.000000   5487.750000   
max    399890.000000       380.000000         89.000000   7000.000000   

       text_message_count         house  handset_price  \
count         9354.000000  9.354000e+03    9354.000000   
mean           135.830874  8.735713e+05     794.461941   
std             48.901110  2.868228e+05    1325.558148   
min             52.000000  3.200010e+05     215.000000   
25%             93.000000  6.390412e+05     497.000000   
50%            136.000000  8.722050e+05     777.000000   
75%            178.000000  1.094315e+06    1065.000000   
max            220.000000  1.456389e+06  125000.000000   

       over_15mins_calls_per_month  average_call_duration            id  
count                  9354.000000            9354.000000   9354.000000  
mean                     10.480864               9.995724  11896.042228  
std                       6.992176               5.167853   6884.912231  
min                       1.000000               1.000000      2.000000  
25%                       4.000000               6.000000   6063.000000  
50%                      10.000000              10.000000  11796.000000  
75%                      17.000000              14.000000  17480.750000  
max                      25.000000              19.000000  25354.000000

## Part 6: Handle Outliers Using Quantiles

Instead of removing outliers, we can limit their impact by capping extreme values — a method known as **Winsorizing**.

### How to Do It:
- Use `.quantile()` to identify the 1st and 99th percentiles (or other thresholds).
- Use `.clip()` to cap values within that range.

This keeps your dataset intact while reducing the influence of extreme values on your analysis or model.



In [ ]:
# Calculate 1st and 99th percentiles for income
income_min, income_max = df['income'].quantile([0.01, 0.99])

# Use .loc to avoid SettingWithCopyWarning and ensure assignment modifies the original DataFrame
df.loc[:, 'income'] = df['income'].clip(lower=income_min, upper=income_max)

# Clip 'data_mb_used' to within 1st and 99th percentiles
usage_min, usage_max = df['data_mb_used'].quantile([0.01, 0.99])
df.loc[:, 'data_mb_used'] = df['data_mb_used'].clip(lower=usage_min, upper=usage_max)

# Clip 'average_call_duration' to reduce the effect of extreme outliers
call_min, call_max = df['average_call_duration'].quantile([0.01, 0.99])
df.loc[:, 'average_call_duration'] = df['average_call_duration'].clip(lower=call_min, upper=call_max)



### 🔧 Try It Yourself – Part 6

1. Use `.quantile([0.01, 0.99])` to find the range for:
  - `text_message_count`
  - `over_15mins_calls_per_month`
2. Apply `.clip(lower=..., upper=...)` to reduce the impact of those outliers
3. Compare the `.describe()` output before and after clipping and comment on what you observe


In [30]:
# Compare the .describe() output before clipping
print("Descriptive statistics before clipping:")
display(df.describe())

# Calculate 1st and 99th percentiles for text_message_count and over_15mins_calls_per_month
text_min, text_max = df['text_message_count'].quantile([0.01, 0.99])
calls_min, calls_max = df['over_15mins_calls_per_month'].quantile([0.01, 0.99])

# Apply .clip() to reduce the impact of outliers using .loc
df.loc[:, 'text_message_count'] = df['text_message_count'].clip(lower=text_min, upper=text_max)
df.loc[:, 'over_15mins_calls_per_month'] = df['over_15mins_calls_per_month'].clip(lower=calls_min, upper=calls_max)

# Compare the .describe() output after clipping
print("\nDescriptive statistics after clipping:")
display(df.describe())

Descriptive statistics before clipping:


income  data_overage_mb  data_leftover_mb  data_mb_used  \
count    9354.000000      9354.000000       9354.000000   9354.000000   
mean   226190.368612       159.193286         37.683985   3990.370002   
std    100519.850725       110.789272         27.859459   1725.193341   
min     52491.000000         0.000000          0.000000   1001.000000   
25%    138562.000000        61.000000         13.000000   2497.000000   
50%    225878.500000       155.000000         35.000000   4044.000000   
75%    313180.000000       246.000000         62.000000   5487.750000   
max    399890.000000       380.000000         89.000000   7000.000000   

       text_message_count         house  handset_price  \
count         9354.000000  9.354000e+03    9354.000000   
mean           135.830554  8.735713e+05     794.461941   
std             48.875726  2.868228e+05    1325.558148   
min             53.000000  3.200010e+05     215.000000   
25%             93.000000  6.390412e+05     497.000000   
50%            136.000000  8.722050e+05     777.000000   
75%            178.000000  1.094315e+06    1065.000000   
max            219.000000  1.456389e+06  125000.000000   

       over_15mins_calls_per_month  average_call_duration            id  
count                  9354.000000            9354.000000   9354.000000  
mean                     10.480864               9.995724  11896.042228  
std                       6.992176               5.167853   6884.912231  
min                       1.000000               1.000000      2.000000  
25%                       4.000000               6.000000   6063.000000  
50%                      10.000000              10.000000  11796.000000  
75%                      17.000000              14.000000  17480.750000  
max                      25.000000              19.000000  25354.000000


Descriptive statistics after clipping:


income  data_overage_mb  data_leftover_mb  data_mb_used  \
count    9354.000000      9354.000000       9354.000000   9354.000000   
mean   226190.368612       159.193286         37.683985   3990.370002   
std    100519.850725       110.789272         27.859459   1725.193341   
min     52491.000000         0.000000          0.000000   1001.000000   
25%    138562.000000        61.000000         13.000000   2497.000000   
50%    225878.500000       155.000000         35.000000   4044.000000   
75%    313180.000000       246.000000         62.000000   5487.750000   
max    399890.000000       380.000000         89.000000   7000.000000   

       text_message_count         house  handset_price  \
count         9354.000000  9.354000e+03    9354.000000   
mean           135.830554  8.735713e+05     794.461941   
std             48.875726  2.868228e+05    1325.558148   
min             53.000000  3.200010e+05     215.000000   
25%             93.000000  6.390412e+05     497.000000   
50%            136.000000  8.722050e+05     777.000000   
75%            178.000000  1.094315e+06    1065.000000   
max            219.000000  1.456389e+06  125000.000000   

       over_15mins_calls_per_month  average_call_duration            id  
count                  9354.000000            9354.000000   9354.000000  
mean                     10.480864               9.995724  11896.042228  
std                       6.992176               5.167853   6884.912231  
min                       1.000000               1.000000      2.000000  
25%                       4.000000               6.000000   6063.000000  
50%                      10.000000              10.000000  11796.000000  
75%                      17.000000              14.000000  17480.750000  
max                      25.000000              19.000000  25354.000000

The statistics before and after clipping stayed the same for text_message_count and over_15mins_calls_per_month, meaning all values already fit within the thresholds. This confirms there were no extreme outliers in those two columns.

## 🔧 Part 7: Reflection (100 words or less per question)

1. Which step fixed the most issues in the dataset?
2. What surprised you about the structure or values?
3. Do you feel this data is now ready for transformation in Lab 7?


1- Handling missing values in part 3 fixed the most issues, as several key columns like income, data_mb_used, and considering_change_of_plan had gaps that could distort analysis. Filling them with median or mode ensured consistency while preserving data volume.

2- I was surprised to find negative and unrealistic values, such as negative income and handset prices. Also, the presence of 3,456 duplicate IDs highlighted that multiple records existed per customer, which required careful handling.

3- Yes, the dataset now has consistent column types, missing values addressed, duplicates removed, and outliers handled. It is clean and reliable, and I think it is ready for transformations in Lab 7.

## Export Your Notebook to Submit in Canvas
- Use the instructions from Lab 1

In [31]:
!jupyter nbconvert --to html "lab_06_AlhinaiAlmuhanna.ipynb"

[NbConvertApp] Converting notebook lab_06_AlhinaiAlmuhanna.ipynb to html
[NbConvertApp] Writing 385174 bytes to lab_06_AlhinaiAlmuhanna.html
